File: autobot-vscode/src/tools.ts
import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import { execSync } from 'child_process';
import { AgentMode, AGENT_MODES } from './config';

interface ToolResult {
    success: boolean;
    result?: string;
    error?: string;
}

interface ToolDef {
    name: string;
    description: string;
    parameters: any;
    handler: (...args: any[]) => Promise<ToolResult>;
    group: string;
}

export class ToolSystem {
    private tools: Map<string, ToolDef> = new Map();

    constructor() {
        this.registerCoreTools();
    }

    getDefinitions(mode: AgentMode): any[] {
        const modeConf = AGENT_MODES[mode];
        const allowed = new Set<string>();
        for (const [group, enabled] of Object.entries(modeConf.groups)) {
            if (enabled) allowed.add(group);
        }

        return Array.from(this.tools.values())
            .filter(t => allowed.has(t.group))
            .map(t => ({
                type: 'function',
                function: {
                    name: t.name,
                    description: t.description,
                    parameters: t.parameters,
                },
            }));
    }

    async execute(name: string, args: any, mode: AgentMode): Promise<ToolResult> {
        const tool = this.tools.get(name);
        if (!tool) return { success: false, error: `Unknown tool: ${name}` };

        const modeConf = AGENT_MODES[mode];
        if (!(modeConf.groups as any)[tool.group]) {
            return { success: false, error: `Tool '${name}' not allowed in ${mode} mode` };
        }

        try {
            return await tool.handler(args);
        } catch (e: any) {
            return { success: false, error: e.message };
        }
    }

    private register(name: string, description: string, parameters: any, handler: (...args: any[]) => Promise<ToolResult>, group: string) {
        this.tools.set(name, { name, description, parameters, handler, group });
    }

    private registerCoreTools() {
        // ── Read tools ──
        this.register('read_file', 'Read the contents of a file',
            { type: 'object', properties: { path: { type: 'string' } }, required: ['path'] },
            async (args) => {
                const filePath = path.resolve(args.path);
                if (!fs.existsSync(filePath)) return { success: false, error: 'File not found' };
                const stat = fs.statSync(filePath);
                if (stat.isDirectory()) return { success: false, error: 'Is a directory' };
                if (stat.size > 500000) return { success: false, error: 'File too large (>500KB)' };
                const content = fs.readFileSync(filePath, 'utf-8');
                return { success: true, result: content };
            }, 'read');

        this.register('search_files', 'Search for files by glob pattern',
            { type: 'object', properties: { pattern: { type: 'string' } }, required: ['pattern'] },
            async (args) => {
                const workspaceFolders = vscode.workspace.workspaceFolders;
                if (!workspaceFolders) return { success: false, error: 'No workspace open' };
                const root = workspaceFolders[0].uri.fsPath;
                const glob = new vscode.RelativePattern(root, args.pattern);
                const files = await vscode.workspace.findFiles(glob, undefined, 200);
                return { success: true, result: files.map(f => f.fsPath).join('\n') };
            }, 'read');

        this.register('grep_search', 'Search file contents using regex',
            { type: 'object', properties: { pattern: { type: 'string' }, path: { type: 'string' } }, required: ['pattern'] },
            async (args) => {
                const workspaceFolders = vscode.workspace.workspaceFolders;
                if (!workspaceFolders) return { success: false, error: 'No workspace open' };
                try {
                    const root = workspaceFolders[0].uri.fsPath;
                    const grepPath = args.path ? path.resolve(root, args.path) : root;
                    const cmd = `grep -rn "${args.pattern}" "${grepPath}" --include="*.py" --include="*.ts" --include="*.js" --include="*.tsx" --include="*.json" --include="*.yaml" --include="*.md" 2>/dev/null | head -100`;
                    const result = execSync(cmd, { cwd: root, timeout: 15000 }).toString();
                    return { success: true, result: result || 'No matches' };
                } catch (e: any) {
                    return { success: false, error: e.message };
                }
            }, 'read');

        this.register('list_files', 'List files in a directory',
            { type: 'object', properties: { path: { type: 'string' }, recursive: { type: 'boolean' } }, required: ['path'] },
            async (args) => {
                const dirPath = path.resolve(args.path);
                if (!fs.existsSync(dirPath)) return { success: false, error: 'Directory not found' };
                if (args.recursive) {
                    const entries: string[] = [];
                    const walkDir = (dir: string, prefix: string = '') => {
                        const items = fs.readdirSync(dir, { withFileTypes: true });
                        for (const item of items) {
                            entries.push(prefix + item.name);
                            if (item.isDirectory()) walkDir(path.join(dir, item.name), prefix + item.name + '/');
                        }
                    };
                    walkDir(dirPath);
                    return { success: true, result: entries.slice(0, 200).join('\n') };
                }
                const items = fs.readdirSync(dirPath);
                return { success: true, result: items.join('\n') };
            }, 'read');

        // ── Edit tools ──
        this.register('write_to_file', 'Write content to a file (creates/overwrites)',
            { type: 'object', properties: { path: { type: 'string' }, content: { type: 'string' } }, required: ['path', 'content'] },
            async (args) => {
                const filePath = path.resolve(args.path);
                fs.mkdirSync(path.dirname(filePath), { recursive: true });
                fs.writeFileSync(filePath, args.content, 'utf-8');
                return { success: true, result: `Wrote ${args.content.length} chars to ${args.path}` };
            }, 'edit');

        this.register('edit_file', 'Find and replace text in a file',
            { type: 'object', properties: { path: { type: 'string' }, old_string: { type: 'string' }, new_string: { type: 'string' } }, required: ['path', 'old_string', 'new_string'] },
            async (args) => {
                const filePath = path.resolve(args.path);
                if (!fs.existsSync(filePath)) return { success: false, error: 'File not found' };
                let content = fs.readFileSync(filePath, 'utf-8');
                if (!content.includes(args.old_string)) return { success: false, error: 'String not found' };
                content = content.replace(args.old_string, args.new_string);
                fs.writeFileSync(filePath, content, 'utf-8');
                return { success: true, result: `Edited ${args.path}` };
            }, 'edit');

        // ── Command tools ──
        this.register('execute_command', 'Run a shell command',
            { type: 'object', properties: { command: { type: 'string' }, timeout: { type: 'number' } }, required: ['command'] },
            async (args) => {
                const workspaceFolders = vscode.workspace.workspaceFolders;
                const cwd = workspaceFolders ? workspaceFolders[0].uri.fsPath : process.cwd();
                try {
                    const result = execSync(args.command, {
                        cwd,
                        timeout: (args.timeout || 30) * 1000,
                        maxBuffer: 1024 * 1024,
                    }).toString();
                    return { success: true, result: result.slice(-10000) || '(no output)' };
                } catch (e: any) {
                    return { success: true, result: e.stdout?.toString().slice(-5000) || e.message };
                }
            }, 'command');

        // ── Learning/Memory tools ──
        this.register('think', 'Think step by step (no side effects)',
            { type: 'object', properties: { thought: { type: 'string' } }, required: ['thought'] },
            async (args) => ({ success: true, result: `Thinking: ${args.thought.slice(0, 200)}...` }), 'read');

        this.register('attempt_completion', 'Signal that the task is complete',
            { type: 'object', properties: { result: { type: 'string' } }, required: ['result'] },
            async (args) => ({ success: true, result: `Task complete: ${args.result}` }), 'read');

        this.register('ask_question', 'Ask the user a question',
            { type: 'object', properties: { question: { type: 'string' } }, required: ['question'] },
            async (args) => {
                const answer = await vscode.window.showInputBox({
                    prompt: args.question,
                    ignoreFocusOut: true,
                });
                return { success: true, result: answer || '(no answer)' };
            }, 'read');

        this.register('open_file', 'Open a file in the editor',
            { type: 'object', properties: { path: { type: 'string' } }, required: ['path'] },
            async (args) => {
                const filePath = path.resolve(args.path);
                const doc = await vscode.workspace.openTextDocument(filePath);
                await vscode.window.showTextDocument(doc);
                return { success: true, result: `Opened ${args.path}` };
            }, 'read');
    }
}

File: autobot/tools.py
import asyncio
import os
import subprocess
from pathlib import Path
from typing import Any, Dict, List, Optional

from .config import Config, MODE_GROUPS


class ToolResult:
    def __init__(self, success: bool, output: str = "", error: str = ""):
        self.success = success
        self.output = output
        self.error = error

    def to_dict(self) -> Dict[str, Any]:
        return {"success": self.success, "output": self.output, "error": self.error}


class ToolRegistry:
    def __init__(self, workspace_root: Optional[str] = None):
        self.workspace_root = Path(workspace_root or os.getcwd()).resolve()
        self.config = Config()
        self._tools: Dict[str, Any] = {}
        self._register_core()

    def _register_core(self):
        self.register("read_file", self._read_file, "read")
        self.register("write_file", self._write_file, "edit")
        self.register("edit_file", self._edit_file, "edit")
        self.register("list_directory", self._list_directory, "read")
        self.register("search_files", self._search_files, "read")
        self.register("execute_command", self._execute_command, "command")
        self.register("think", self._think, "read")
        self.register("attempt_completion", self._attempt_completion, "read")

    def register(self, name: str, handler: Any, group: str):
        self._tools[name] = {"name": name, "handler": handler, "group": group}

    def get_definitions(self, mode: str) -> List[Dict[str, Any]]:
        allowed_groups = MODE_GROUPS.get(mode, MODE_GROUPS["coder"])
        allowed = {g for g, ok in allowed_groups.__dict__.items() if ok}
        definitions = []
        for tool in self._tools.values():
            if tool["group"] in allowed:
                definitions.append(self._schema(tool["name"]))
        return definitions

    def _schema(self, name: str) -> Dict[str, Any]:
        schemas = {
            "read_file": {
                "type": "function",
                "function": {
                    "name": "read_file",
                    "description": "Read the contents of a file",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "path": {"type": "string", "description": "Relative or absolute path to the file"}
                        },
                        "required": ["path"],
                    },
                },
            },
            "write_file": {
                "type": "function",
                "function": {
                    "name": "write_file",
                    "description": "Write content to a file (creates or overwrites)",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "path": {"type": "string"},
                            "content": {"type": "string"},
                        },
                        "required": ["path", "content"],
                    },
                },
            },
            "edit_file": {
                "type": "function",
                "function": {
                    "name": "edit_file",
                    "description": "Find and replace text in a file",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "path": {"type": "string"},
                            "old_string": {"type": "string"},
                            "new_string": {"type": "string"},
                        },
                        "required": ["path", "old_string", "new_string"],
                    },
                },
            },
            "list_directory": {
                "type": "function",
                "function": {
                    "name": "list_directory",
                    "description": "List files in a directory",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "path": {"type": "string"},
                            "recursive": {"type": "boolean", "default": False},
                        },
                        "required": ["path"],
                    },
                },
            },
            "search_files": {
                "type": "function",
                "function": {
                    "name": "search_files",
                    "description": "Search file contents using regex",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "pattern": {"type": "string"},
                            "path": {"type": "string", "default": "."},
                        },
                        "required": ["pattern"],
                    },
                },
            },
            "execute_command": {
                "type": "function",
                "function": {
                    "name": "execute_command",
                    "description": "Run a shell command in the workspace",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "command": {"type": "string"},
                            "timeout": {"type": "number", "default": 30},
                            "cwd": {"type": "string", "default": "."},
                        },
                        "required": ["command"],
                    },
                },
            },
            "think": {
                "type": "function",
                "function": {
                    "name": "think",
                    "description": "Think step by step (no side effects)",
                    "parameters": {
                        "type": "object",
                        "properties": {"thought": {"type": "string"}},
                        "required": ["thought"],
                    },
                },
            },
            "attempt_completion": {
                "type": "function",
                "function": {
                    "name": "attempt_completion",
                    "description": "Signal that the task is complete",
                    "parameters": {
                        "type": "object",
                        "properties": {"result": {"type": "string"}},
                        "required": ["result"],
                    },
                },
            },
        }
        return schemas.get(name, {"type": "function", "function": {"name": name, "description": "", "parameters": {}}})

    async def execute(self, name: str, arguments: Dict[str, Any], mode: str) -> ToolResult:
        if name not in self._tools:
            return ToolResult(False, error=f"Unknown tool: {name}")

        tool = self._tools[name]
        allowed_groups = MODE_GROUPS.get(mode, MODE_GROUPS["coder"])
        if not getattr(allowed_groups, tool["group"], False):
            return ToolResult(False, error=f"Tool '{name}' not allowed in {mode} mode")

        try:
            result = await tool["handler"](arguments)
            if asyncio.iscoroutine(result):
                result = await result
            if isinstance(result, ToolResult):
                return result
            return ToolResult(True, output=str(result))
        except Exception as e:
            return ToolResult(False, error=str(e))

    def _resolve(self, path: str) -> Path:
        p = Path(path)
        if p.is_absolute():
            return p
        return (self.workspace_root / p).resolve()

    async def _read_file(self, args: Dict[str, Any]) -> ToolResult:
        target = self._resolve(args["path"])
        if not target.exists():
            return ToolResult(False, error=f"File not found: {args['path']}")
        if target.is_dir():
            return ToolResult(False, error="Is a directory")
        if target.stat().st_size > 2_000_000:
            return ToolResult(False, error="File too large (>2MB)")
        content = target.read_text(encoding="utf-8", errors="replace")
        return ToolResult(True, output=content)

    async def _write_file(self, args: Dict[str, Any]) -> ToolResult:
        target = self._resolve(args["path"])
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(args["content"], encoding="utf-8")
        return ToolResult(True, output=f"Wrote {len(args['content'])} chars to {args['path']}")

    async def _edit_file(self, args: Dict[str, Any]) -> ToolResult:
        target = self._resolve(args["path"])
        if not target.exists():
            return ToolResult(False, error=f"File not found: {args['path']}")
        content = target.read_text(encoding="utf-8", errors="replace")
        old = args["old_string"]
        new = args["new_string"]
        if old not in content:
            return ToolResult(False, error="old_string not found in file")
        new_content = content.replace(old, new, 1)
        target.write_text(new_content, encoding="utf-8")
        return ToolResult(True, output=f"Edited {args['path']}")

    async def _list_directory(self, args: Dict[str, Any]) -> ToolResult:
        target = self._resolve(args["path"])
        if not target.exists():
            return ToolResult(False, error=f"Directory not found: {args['path']}")
        if not target.is_dir():
            return ToolResult(False, error="Not a directory")
        recursive = args.get("recursive", False)
        entries: List[str] = []

        def walk(dir_path: Path, prefix: str = ""):
            for item in sorted(dir_path.iterdir(), key=lambda x: (x.is_file(), x.name)):
                entries.append(f"{prefix}{item.name}")
                if item.is_dir() and recursive:
                    walk(item, prefix + item.name + "/")

        if recursive:
            walk(target)
        else:
            for item in sorted(target.iterdir(), key=lambda x: (x.is_file(), x.name)):
                entries.append(item.name + ("/" if item.is_dir() else ""))
        return ToolResult(True, output="\n".join(entries[:300]))

    async def _search_files(self, args: Dict[str, Any]) -> ToolResult:
        import re
        pattern = args["pattern"]
        root = self._resolve(args.get("path", "."))
        if not root.exists():
            return ToolResult(False, error=f"Path not found: {args.get('path', '.')}")
        matches = []
        try:
            regex = re.compile(pattern)
        except re.error as e:
            return ToolResult(False, error=f"Invalid regex: {e}")
        for file in root.rglob("*"):
            if file.is_file() and file.suffix.lower() in {
                ".py", ".ts", ".js", ".tsx", ".json", ".yaml", ".yml", ".md", ".txt", ".toml", ".cfg", ".ini"
            }:
                try:
                    text = file.read_text(encoding="utf-8", errors="ignore")
                    for i, line in enumerate(text.splitlines(), 1):
                        if regex.search(line):
                            matches.append(f"{file.relative_to(root)}:{i}: {line.strip()}")
                            if len(matches) >= 100:
                                break
                except Exception:
                    continue
            if len(matches) >= 100:
                break
        return ToolResult(True, output="\n".join(matches) if matches else "No matches")

    async def _execute_command(self, args: Dict[str, Any]) -> ToolResult:
        command = args["command"]
        timeout = args.get("timeout", Config().agent.tool_timeout)
        cwd = args.get("cwd", ".")
        cwd_path = self._resolve(cwd)
        try:
            result = subprocess.run(
                command,
                shell=True,
                cwd=cwd_path,
                capture_output=True,
                text=True,
                timeout=timeout,
                env={**os.environ},
            )
            output = result.stdout.strip()
            if result.stderr.strip():
                output += "\n" + result.stderr.strip()
            success = result.returncode == 0
            return ToolResult(success=success, output=output[:20000], error="" if success else f"Exit code {result.returncode}")
        except subprocess.TimeoutExpired:
            return ToolResult(False, error=f"Command timed out after {timeout}s")
        except Exception as e:
            return ToolResult(False, error=str(e))

    async def _think(self, args: Dict[str, Any]) -> ToolResult:
        return ToolResult(True, output=f"Thought recorded: {args['thought'][:500]}")

    async def _attempt_completion(self, args: Dict[str, Any]) -> ToolResult:
        return ToolResult(True, output=args["result"], error="")
